modules/gateway/upnp.go

```diff
@@ -109,15 +109,22 @@ func (g *Gateway) threadedForwardPort(port string) {
 	}
 
 	g.log.Println("INFO: successfully forwarded port", port)
+
+	// Establish port-clearing at shutdown.
+	g.threads.AfterStop(func() {
+		g.managedClearPort(g.myAddr.Port())
+	})
 }
 
-// clearPort removes a port mapping from the router.
-func (g *Gateway) clearPort(port string) {
+// managedClearPort removes a port mapping from the router.
+func (g *Gateway) managedClearPort(port string) {
```

Comment:
It doesn't need a lock, which is equivalent to it doing it's own locking. Functionally, 'mamaged' tells you not to call it holding a lock. It can take a while too, so you really don't want to be holding a lock